### PR TITLE
helper functions to bootstrap from renv.lock

### PR DESCRIPTION
**Pull request by RichardJActon**
Partially addresses #5, this implements a "fast" inexact conversion from renv to nix environment description. It does **NOT** capture the exact package versions, just the R version and the list of packages in `renv.lock`, nor does it differentiate primary from transitive dependencies in `renv.lock` it simply lists all dependencies.

This PR adds four functions:

- user facing
    - `renv2nix()` the main addition, given the path to an `renv.lock` file this function extracts the list of packages used, including handling packages from github/gitlab remotes, and makes a call to `rix()` with this information.
    - `renv_lock_r_ver()` a utility function for getting the R version recorded in a `renv.lock`
- internal
    - `read_renv_lock()` reads an `renv.lock` file with jsonlite
    - `renv_remote_pkg()` extracts remote information needed for `git_pkgs` from `renv.lock` package entries 

# TODO

- [x] additional testing required - need to check more unexpected inputs in automated testing

### CODE REVIEW

**Review by b-rodrigues — CHANGES_REQUESTED** [review]
Many thanks for this PR! This is really looking great and will be useful to many people!

There are some  things to change, but otherwise, looking good!

File: R/renv_helpers.R [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+    }
+  )
+  pkg_info
+}
+
+#' renv2nix
+#'
+#' @param renv_lock_path location of the renv.lock file, defaults to "renv.lock"
+#' @param return_rix_call return the generated rix function call instead of
+#' evaluating it this is for debugging purposes, defaults to FALSE
```
[the remark this answers is not in the document]
Comment (RichardJActon) [reply R/renv_helpers.R:79]: addressed in [f717e9a](https://github.com/ropensci/rix/pull/349/commits/f717e9a4d134033acbc64fa8e86176eae752f15a)

File: R/renv_helpers.R [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+#' @param ... any other parameters to pass to [rix]
+#'
+#' @return nothing side effects only
```
[the remark this answers is not in the document]
Comment (RichardJActon) [reply R/renv_helpers.R]: addressed in [e7a5d08](https://github.com/ropensci/rix/pull/349/commits/e7a5d0841afe5cadcbf4ab9a827c169967673d14)

File: R/renv_helpers.R [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+#' @export
+#'
+renv2nix <- function(
+    renv_lock_path = "renv.lock",
+    return_rix_call = FALSE,
+    method = "fast",
```
[the remark this answers is not in the document]
Comment (RichardJActon) [reply R/renv_helpers.R]: addressed in [f717e9a](https://github.com/ropensci/rix/pull/349/commits/f717e9a4d134033acbc64fa8e86176eae752f15a)

File: NAMESPACE [hunk 4]
```diff
@@ -3,6 +3,8 @@
 export(available_r)
 export(ga_cachix)
 export(nix_build)
+export(renv2nix)
+export(renv_lock_r_ver)
```
[the remark this answers is not in the document]
Comment (RichardJActon) [reply NAMESPACE:7]: I've found it useful to be able to just check the R version used by a project to gauge roughly how old the R version is that the project uses - especially old ones have in my experience been a bit more complex to debug if you need to fix and CRAN snapshot in renv for a particular date to get all the package versions recorded for instance. If it is not an R version I've used for any other projects it lets me know I'm going to be installing another R version, which might take little longer to build. 

I implemented it before taking the `renv2nix()` function approach so don't use it for anything now but thought it might be a nice convenience function, but happy to drop it if you don't think this is the place for it.

File: R/renv_helpers.R [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+    type = renv_lock_pkg_info$RemoteType) {
+  type <- match.arg(type, c(
+    "github", "gitlab"
+    # , "bitbucket", "git", "local", "svn", "url", "version", "cran", "bioc"
+  ))
+  pkg_info <- vector(mode = "list", length = 3)
+  names(pkg_info) <- c("package_name", "repo_url", "commit")
+  switch(type,
+    "github" = {
+      pkg_info[[1]] <- renv_lock_pkg_info$Package
```
[the remark this answers is not in the document]
Comment (RichardJActon) [reply R/renv_helpers.R:54]: I'm not sure how this would help reduce the repetition. The two blocks differ as github does not use `RemoteHost` and I anticipate that the different remote types may similarly need slightly different handling which is why I repeated the `Package` and `RemoteSha` assignments rather than placing them outside the switch in case one other other remotes need some different processing to get at these values as github does with `RemoteHost`.

File: NAMESPACE (hunk 4 above)
[the remark this answers is not in the document]
Comment (b-rodrigues) [reply NAMESPACE:7]: I’d see we keep this function private.

File: R/renv_helpers.R (hunk 5 above)
[the remark this answers is not in the document]
Comment (b-rodrigues) [reply R/renv_helpers.R:54]: True, but I would prefer this logic to be handled in one function, and there would still be a lot of boilerplate that could go into that function as well.

[the remark this answers is not in the document]
Comment (RichardJActon) [reply R/renv_helpers.R:54]: Do you mean a function which returns a list that can be handed directly to the `rix()` `git_pkgs` argument - like this: https://github.com/RichardJActon/rix/commit/68ebf8adcf8f9568942b0552f31558b1a66fcff4 because I don't really see what can be abstracted out of `renv_remote_pkg()` another function would still need the individual package info, the type, the validation of the type and the logic to change behavior based on the type and that's all that's in `renv_remote_pkg()`?

File: NAMESPACE (hunk 4 above)
[the remark this answers is not in the document]
Comment (RichardJActon) [reply NAMESPACE]: addressed in [0f9de72](https://github.com/ropensci/rix/pull/349/commits/0f9de7290325be6beedde83ac6fa4c46c3487522)

File: R/renv_helpers.R (hunk 5 above)
[the remark this answers is not in the document]
Comment (b-rodrigues) [reply R/renv_helpers.R:54]: Right, the idea is to have all this generated in one place by one function

**Comment by b-rodrigues** [comment]
@philipp-baumann when you'll be have some spare time could you take a look at the PR? It's looking good for me

File: R/renv_helpers.R (hunk 5 above)
[the remark this answers is not in the document]
Comment (RichardJActon) [reply R/renv_helpers.R]: I've consolidated this into one function and added some error handling, and some more tests.

File: tests/testthat/test-renv_helpers.R [hunk 6]
```diff
@@ -116,6 +116,14 @@ testthat::test_that("testing renv_helpers", {
     })
     testthat::expect_equal(call, test_call)
     testthat::expect_match(warns, "has the unsupported remote type")
+
+    warns <- testthat::capture_warnings({
+      call <- renv2nix(tmpf, return_rix_call = TRUE, ide = "rstudio")
+    })
+    test_call$ide <- "rstudio"
+    testthat::expect_equal(call, test_call)
+    testthat::expect_match(warns, "has the unsupported remote type")
```
Comment (b-rodrigues) [inline tests/testthat/test-renv_helpers.R:125]: why not expect_warning?

https://testthat.r-lib.org/reference/expect_error.html

Comment (RichardJActon) [reply tests/testthat/test-renv_helpers.R]: It's a pattern I got into working on another project where I'm regularly testing things that throw multiple warnings, `capture_warnings()` let's me test I've got the right set of warnings. In this case `expect_warning()` is indeed better.

**Comment by RichardJActon** [comment]
Thanks @b-rodrigues I think it is reviewable as is.

In playing with it a little myself I wonder if it might be a good idea to duplicate the arguments from `rix` explicitly rather than passing them with `...` as I find myself missing the auto-completion that you get with explicit arguments this doesn't seem to work with dots even when using inherit dot params in the documentation (at least in Rstudio).

It may be worth taking a look at {renv}'s tests to see what they do, but maybe I should include some more 'real world' `renv.lock` files to check for any unanticipated edge cases my artificial examples haven't covered.

Should have some time this weekend to make tweaks, will likely be busier next weekend and week of the 4th

**Comment by b-rodrigues** [comment]
> maybe I should include some more 'real world' `renv.lock` files to check for any unanticipated edge cases my artificial examples haven't covered.


Yes, this would be great

**Comment by b-rodrigues** [comment]
> In playing with it a little myself I wonder if it might be a good idea to duplicate the arguments from `rix` explicitly rather than passing them with `...`

I think that we should at least provide the `project_path` as being the same folder containing the `renv.lock` file by default, but I would prefer to keep the `...` otherwise (sorry about autocompletion :smile_cat: )

**Comment by b-rodrigues** [comment]
Also, we should not include `renv` in the resulting `default.nix` as it wouldn't be needed, no?

**Comment by b-rodrigues** [comment]
fuuuuuuu

I messed up, sorry! Didn't want to close this, just wanted to add a change...

@RichardJActon sorry, can you re-open the PR

**Comment by b-rodrigues** [comment]
All your changes are in this branch by the way:

https://github.com/ropensci/rix/tree/add_renv2nix

because I think that I somehow overwrote the history of your main branch   :fearful: :exploding_head:
